File: dep/moves.py
from itertools import permutations
import numpy as np

from config import *

class Movement:
    def __init__(self, pawn2step, board_size=None):
        self.moves, self.attacks = [], []
        self.pawn2step = pawn2step
        self.board_size = BOARD_SIZE if board_size is None else board_size

    def check_square(self, player, array, position, only_move=False, only_eat=False):
        # Check that the square is not outside
        for dim in position:
            if dim < 0: return
        for dim, dim_max in zip(position, self.board_size):
            if dim >= dim_max: return

        # If it's empty, we can move to it
        if array[position] == 0:
            if only_eat is False:
                self.moves.append(position)
        # If it's an enemy, we can attack it
        elif player * array[position] < 0:
            if only_move is False:
                self.attacks.append(position)

    def check_line(self, player, array, position, move_x, move_y, move_z):
        # move_x, move_y, move_z can be 1, -1, 0
        # 1 if forward, -1 if backwards 0 if no motion in that dimension
        i = 0

        while True:
            i += 1
            sq = Square(x=position.x + i*move_x, y = position.y + i*move_y, z = position.z + i*move_z)

            # Break if we go outside the board
            for dim in sq:
                if dim < 0: return
            for dim, dim_max in zip(sq, self.board_size):
                if dim >= dim_max: return

            # If it's empty we can move to it
            if array[sq] == 0:
                self.moves.append(sq)
            # If there's an enemy we can attack it but nothing behind it
            elif player * array[sq] < 0:
                self.attacks.append(sq)
                return
            # If there is an ally, we can't go through them
            else:
                return

    def allowed_moves(self, player, array, position):
        '''Returns the moves and attacks a piece can perform'''
        piece = abs(array[position])
        if piece == 1: return self.pawn_moves(player, array, position)
        if piece == 2: return self.rook_moves(player, array, position)
        if piece == 3: return self.knight_moves(player, array, position)
        if piece == 4: return self.bishop_moves(player, array, position)
        if piece == 5: return self.queen_moves(player, array, position)
        if piece == 6: return self.king_moves(player, array, position)
        if piece == 7: return self.unicorn_moves(player, array, position)

    def pawn_moves(self, player, array, position):
        x, y, z = position
        # They can only move forwards (whites upwards, blacks downwards)
        self.check_square(player, array, Square(x, y+1*player, z), only_move=True)

        # If we are in the second row we can move twice (in certain gamemodes)
        if self.pawn2step is True:
            if (player == 1 and y == 1) or (player == -1 and y == self.board_size.y-2):
                # If there's nothing on the first one
                if array[x, y+1*player, z] == 0:
                    self.check_square(player, array, Square(x, y+2*player, z), only_move=True)

        self.check_square(player, array, Square(x, y, z+1*player), only_move=True)

        self.check_square(player, array, Square(x+1, y+1*player, z), only_eat=True)
        self.check_square(player, array, Square(x-1, y+1*player, z), only_eat=True)

        self.check_square(player, array, Square(x+1, y, z+1*player), only_eat=True)
        self.check_square(player, array, Square(x-1, y, z+1*player), only_eat=True)

        self.check_square(player, array, Square(x, y+1*player, z+1*player), only_eat=True)
        
        return self.moves, self.attacks
        
    def rook_moves(self, player, array, position):
        # Can move in straight lines in any dimension
        for d in (-1, 1):
            self.check_line(player, array, position, d,0,0)
            self.check_line(player, array, position, 0,d,0)
            self.check_line(player, array, position, 0,0,d)
        return self.moves, self.attacks

    def knight_moves(self, player, array, position):
        # They can move in any permutation of 2 steps one dimension, 1 step in another dimension
        perms  = list(permutations((1,2,0)))
        perms += list(permutations((1,-2,0)))
        perms += list(permutations((-1,2,0)))
        perms += list(permutations((-1,-2,0)))
        perms = list(set(perms))

        for perm in perms:
            x, y, z = [dim+dim2 for dim, dim2 in zip(position, perm)]
            self.check_square(player, array, Square(x, y, z))
        return self.moves, self.attacks

    def bishop_moves(self, player, array, position):
        # They can move as normal bishops
        self.check_line(player, array, position, 1, 1, 0)
        self.check_line(player, array, position, -1, 1, 0)
        self.check_line(player, array, position, 1, -1, 0)
        self.check_line(player, array, position, -1, -1, 0)

        # And also in a cross through the Z dimension
        self.check_line(player, array, position, 1, 0, 1)
        self.check_line(player, array, position, -1, 0, -1)
        self.check_line(player, array, position, 0, 1, 1)
        self.check_line(player, array, position, 0, -1, -1)
        return self.moves, self.attacks

    def queen_moves(self, player, array, position):
        # Combines rook, bishop and unicorn
        # Rook
        for d in (-1, 1):
            self.check_line(player, array, position, d,0,0)
            self.check_line(player, array, position, 0,d,0)
            self.check_line(player, array, position, 0,0,d)

        # Bishop
        self.check_line(player, array, position, 1, 1, 0)
        self.check_line(player, array, position, -1, 1, 0)
        self.check_line(player, array, position, 1, -1, 0)
        self.check_line(player, array, position, -1, -1, 0)

        self.check_line(player, array, position, 1, 0, 1)
        self.check_line(player, array, position, -1, 0, -1)
        self.check_line(player, array, position, 0, 1, 1)
        self.check_line(player, array, position, 0, -1, -1)

        # Unicorn
        for z in (-1, 1):
            self.check_line(player, array, position, 1,1,z)
            self.check_line(player, array, position, -1,1,z)
            self.check_line(player, array, position, 1,-1,z)
            self.check_line(player, array, position, -1,-1,z)
        return self.moves, self.attacks

    def king_moves(self, player, array, position):
        # Moves 1 step in any direction
        perms  = list(permutations((1,1,0)))
        perms += list(permutations((-1,1,0)))
        # perms += list(permutations((1,-1,0)))
        perms += list(permutations((-1,-1,0)))

        perms += list(permutations((0,1,0)))
        perms += list(permutations((-1,0,0)))
        perms = list(set(perms))

        for perm in perms:
            x, y, z = [dim+dim2 for dim, dim2 in zip(position, perm)]
            self.check_square(player, array, Square(x, y, z))
        return self.moves, self.attacks
    
    def unicorn_moves(self, player, array, position):
        # Moves in diagonals through the Z dimension
        for z in (-1, 1):
            self.check_line(player, array, position, 1,1,z)
            self.check_line(player, array, position, -1,1,z)
            self.check_line(player, array, position, 1,-1,z)
            self.check_line(player, array, position, -1,-1,z)
        return self.moves, self.attacks

File: ray/Moves.cpp
#include <iostream>
#include <string>
#include <vector>
#include <vector>
#include "AI.h"
#include "Moves.h"

namespace moves {

// Just constructors
Moves::Moves() {}

Moves::Moves(chess::Board* t_board, int t_x, int t_y, int t_z)
: m_board(t_board), m_x(t_x), m_y(t_y), m_z(t_z) {}

Moves::~Moves() {}

bool Moves::getCheck() {
  return m_check;
}

// Check if we can move to a square
bool Moves::checkSquare(int t_x, int t_y, int t_z, bool can_attack, bool can_move) {
  // If the square is outside the chessboard, return false
  if (t_x<0 || t_y<0 || t_z<0 || t_x>=m_board->getMaxX() || t_y>=m_board->getMaxY() || t_z>=m_board->getMaxZ()) {
    return false;
  }

  // If there is an enemy (and we can attack with this move)
  if (m_board->getTurn()*m_board->get(t_x,t_y,t_z) < 0 && can_attack==true) {
    m_possibleMoves.push_back({std::vector<int> {m_x, m_y, m_z}, std::vector<int> {t_x,t_y,t_z}});

    // If the enemy is a king, we set CHECK to TRUE
    if (abs(m_board->get(t_x,t_y,t_z))==6) {
      m_check = true;
    }
    // Return false because we can't keep moving after attacking this piece
    return false;
  }
  // If it's empty (and we are allowed to move to this position)
  if (m_board->get(t_x,t_y,t_z) == 0 && can_move==true) {
    m_possibleMoves.push_back({std::vector<int> {m_x, m_y, m_z}, std::vector<int> {t_x,t_y,t_z}});
    return true;
  }
  return false;
}

// Check how far we can move through a line
void Moves::checkLine(int t_x, int t_y, int t_z, int t_moveX, int t_moveY, int t_moveZ) {
  // # move_x, move_y, move_z can be 1, -1, 0
  // # 1 if forward, -1 if backwards 0 if no motion in that dimension
  int i = 0;

  while (1) {
    i++;
    int goX = t_x+i*t_moveX;
    int goY = t_y+i*t_moveY;
    int goZ = t_z+i*t_moveZ;
    // If we couldn't go to this square, we can't go behind it either so just quit
    if (checkSquare(goX, goY, goZ) == false) {
      return;
    }
  }
}



std::vector<std::pair<std::vector<int>,std::vector<int>>> Moves::pawnMoves() {
  int turn = (*m_board).getTurn();

  // MOVING FORWARD (CAN'T ATTACK ON THIS MOVE)
  if (checkSquare(m_x, m_y+1*turn, m_z, false)) {
    if (m_board->getPawn2step()) {
      // If we are on the starting pawn position
      // (one step away from border and on the lowest(white)/highest(black) layer of the board)
      if (( turn == 1 && m_y == 1 && m_z == 0 )
      || ( turn == -1 && m_y == m_board->getMaxY()-2
      && m_z == m_board->getMaxZ()-1 )) {
        checkSquare(m_x, m_y+2*turn, m_z, false);
      }
    }
  }

  // Move upwards
  checkSquare(m_x, m_y, m_z+1*turn);
  

  // DIAGONALS (CAN ONLY ATTACK ON THESE)
  checkSquare(m_x+1, m_y+1*turn, m_z, true, false);
  checkSquare(m_x-1, m_y+1*turn, m_z, true, false);

  checkSquare(m_x+1, m_y, m_z+1*turn, true, false);
  checkSquare(m_x-1, m_y, m_z+1*turn, true, false);

  checkSquare(m_x, m_y+1*turn, m_z+1*turn, true, false);

  return m_possibleMoves;
}





std::vector<std::pair<std::vector<int>,std::vector<int>>> Moves::rookMoves() {
  // Straight lines in every dimension
  for (auto d : {-1, 1}) {
    checkLine(m_x,m_y,m_z, d,0,0);
    checkLine(m_x,m_y,m_z, 0,d,0);
    checkLine(m_x,m_y,m_z, 0,0,d);
  }
  return m_possibleMoves;
}

std::vector<std::pair<std::vector<int>,std::vector<int>>> Moves::knightMoves() {
  for (auto d : {-1, 1}) {
    for (auto d2 : {-2, 2}) {
      // Diagonal steps in x,y axis
      checkSquare(m_x+d, m_y+d2, m_z);
      checkSquare(m_x+d2, m_y+d, m_z);
      // Diagonal steps in x,z axis
      checkSquare(m_x+d, m_y, m_z+d2);
      checkSquare(m_x+d2, m_y, m_z+d);
      // Diagonal steps in y,z axis
      checkSquare(m_x, m_y+d, m_z+d2);
      checkSquare(m_x, m_y+d2, m_z+d);
    }
  }
  return m_possibleMoves;
}



std::vector<std::pair<std::vector<int>,std::vector<int>>> Moves::bishopMoves() {
  for (auto d : {-1, 1}) {
    // Diagonals like a normal bishop
    checkLine(m_x, m_y, m_z, d,d*-1,0);
    checkLine(m_x, m_y, m_z, d,d,0);
    // And straight lines through the Z dimension
    checkLine(m_x, m_y, m_z, d,0,d);
    checkLine(m_x, m_y, m_z, 0,d,d);
  }
  return m_possibleMoves;
}

std::vector<std::pair<std::vector<int>,std::vector<int>>> Moves::queenMoves() {
  // Rook+Bishop+Unicorn
  for (auto d : {-1, 1}) {
    // Rook moves
    checkLine(m_x, m_y, m_z, d,0,0);
    checkLine(m_x, m_y, m_z, 0,d,0);
    checkLine(m_x, m_y, m_z, 0,0,d);

    // Bishop moves
    checkLine(m_x, m_y, m_z, d,d*-1,0);
    checkLine(m_x, m_y, m_z, d,d,0);
    checkLine(m_x, m_y, m_z, d,0,d);
    checkLine(m_x, m_y, m_z, 0,d,d);

    // Unicorn moves
    checkLine(m_x, m_y, m_z, 1,1,d);
    checkLine(m_x, m_y, m_z, -1,1,d);
    checkLine(m_x, m_y, m_z, 1,-1,d);
    checkLine(m_x, m_y, m_z, -1,-1,d);
  }
  return m_possibleMoves;
}

std::vector<std::pair<std::vector<int>,std::vector<int>>> Moves::kingMoves() {
  for (auto d : {-1, 1}) {
    // 1 Step in every direction
    checkSquare(m_x+d, m_y, m_z);
    checkSquare(m_x, m_y+d, m_z);
    checkSquare(m_x, m_y, m_z+d);
    for (auto d2 : {-1, 1}) {
      // Diagonal steps in x,y axis
      checkSquare(m_x+d, m_y+d2, m_z);
      // Diagonal steps in x,z axis
      checkSquare(m_x+d, m_y, m_z+d2);
      // Diagonal steps in y,z axis
      checkSquare(m_x, m_y+d, m_z+d2);
      for (auto d3 : {-1, 1}) {
        // Superdiagonal steps (x,y,z)
        checkSquare(m_x+d, m_y+d2, m_z+d3);
      }
    }
  }
  return m_possibleMoves;
}


std::vector<std::pair<std::vector<int>,std::vector<int>>> Moves::unicornMoves() {
  for (auto d : {-1, 1}) {
    // Diagonals through the Z dimension
    checkLine(m_x, m_y, m_z, 1,1,d);
    checkLine(m_x, m_y, m_z, -1,1,d);
    checkLine(m_x, m_y, m_z, 1,-1,d);
    checkLine(m_x, m_y, m_z, -1,-1,d);
  }
  return m_possibleMoves;
}


}
